core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -97,11 +101,17 @@ case class ExceptionFailure(
     description: String,
     stackTrace: Array[StackTraceElement],
     fullStackTrace: String,
-    metrics: Option[TaskMetrics])
+    metrics: Option[TaskMetrics],
+    exception: Option[Throwable] = None)
```

Comment:
I think we're definitely out of luck for binary compatibility, but I think @pwendell just wanted to preserve source compatibility (ie. maybe users will need to recompile, but they won't need to change their code at all).

However, I don't think that is possible either.  You would need to have another method like `def unapply(ef: ExceptionFailure): Option[(String, String, Array[StackTraceElement], String, Option[TaskMetrics])]` -- ie., exactly the same as the built-in unapply, but without the final `Option[Throwable]` in the return type.  But that isn't legal overloading -- it has the same set of arguments as the built-in `unapply`, just a different return type.

Is there another way around this I'm not seeing?  I agree we shouldn't change things willy-nilly just b/c its `@DeveloperApi`, but IMO this change is worth it.